File: js/tests/Comments.test.js
// Conformance tests for line comments (issue #301).
//
// `#` starts a comment that runs to the end of the line, unless it sits inside
// a token or inside a delimited reference. Parsers accept comments by default
// and can be told to treat `#` as an ordinary character again. The table below
// is shared with the Rust, Python, C#, Go, Java and PHP suites, so a document
// written by one implementation reads the same in all of them.

import { test, expect } from 'bun:test';
import { Parser, ParseError, stripComments } from '../src/index.js';

const parser = new Parser();

// Render a parsed node unambiguously: every reference is wrapped in angle
// brackets, so a `#` that survived as content is visible.
function render(node) {
  if (!node.values || node.values.length === 0) {
    return `<${node.id ?? ''}>`;
  }
  const head =
    node.id === null || node.id === undefined ? '' : `<${node.id}>: `;
  return `(${head}${node.values.map(render).join(' ')})`;
}

function parsesAs(input) {
  return parser.parse(input).map(render).join('\n');
}

test('a line that starts with a hash is a comment', () => {
  expect(parsesAs('# a b\n')).toBe('');
});

test('a comment may hold a colon', () => {
  // The document from #301: prose with a colon used to be a parse error.
  expect(parsesAs('# a: b\n')).toBe('');
});

test('a comment may hold anything at all', () => {
  expect(parsesAs('# ) : ( " \' ` #\n')).toBe('');
});

test('a comment ends at the end of its line', () => {
  expect(parsesAs('# note\na: b\n')).toBe('(<a>: <b>)');
});

test('a comment may follow a link', () => {
  expect(parsesAs('a: b # why\n')).toBe('(<a>: <b>)');
});

test('a comment may follow a group', () => {
  expect(parsesAs('(a b) # why\n')).toBe('(<a> <b>)');
});

test('a comment needs no closing newline', () => {
  expect(parsesAs('a: b # why')).toBe('(<a>: <b>)');
});

test('a comment line inside an indented block is skipped', () => {
  expect(parsesAs('parent\n  # what the child is for\n  child\n')).toBe(
    parsesAs('parent\n  child\n')
  );
});

test('a comment line inside a group is skipped', () => {
  expect(parsesAs('(\n  a\n  # why\n  b\n)\n')).toBe(
    parsesAs('(\n  a\n  b\n)\n')
  );
});

test('a line of spaces separates links the way an empty line does', () => {
  // Blanking a comment leaves a line of spaces behind, so such a line has to
  // read as a blank line.
  expect(parsesAs('a\n   \nb\n')).toBe(parsesAs('a\n\nb\n'));
});

test('a document of comments alone holds no links', () => {
  expect(parsesAs('# one\n# two\n')).toBe('');
});

test('a hash inside a token is an ordinary character', () => {
  expect(parsesAs('issue#1047\n')).toBe('(<issue#1047>)');
});

test('a hash that opens a token is an ordinary character', () => {
  expect(parsesAs('a: b#c\n')).toBe('(<a>: <b#c>)');
});

test('a hash inside a delimited reference is content', () => {
  expect(parsesAs('"# not a comment" a\n')).toBe('(<# not a comment> <a>)');
});

test('a comment may follow a delimited reference', () => {
  expect(parsesAs('"a" # why\n')).toBe('(<a>)');
});

test('a hash inside a multiline delimited reference is content', () => {
  expect(parsesAs('"a # b\nc" d\n')).toBe('(<a # b\nc> <d>)');
});

test('comments can be turned off', () => {
  const plain = new Parser({ comments: false });

  expect(plain.parse('# a b\n').map(render).join('\n')).toBe('(<#> <a> <b>)');
});

test('a parser without comments still rejects the document from the issue', () => {
  const plain = new Parser({ comments: false });

  expect(() => plain.parse('# a: b\n')).toThrow(ParseError);
});

test('comments are on by default', () => {
  expect(new Parser().comments).toBe(true);
  expect(new Parser({}).comments).toBe(true);
});

test('a comment does not move the position a later error is reported at', () => {
  // Blanking a comment keeps every later character where it was, so the
  // position reported for a defect is the position in the original document.
  let error = null;
  try {
    parser.parse('# a comment\nstage: rust: nextest\n');
  } catch (thrown) {
    error = thrown;
  }

  expect(error).toBeInstanceOf(ParseError);
  expect(error.line).toBe(2);
  expect(error.column).toBe(12);
  expect(error.lineText).toBe('stage: rust: nextest');
});

test('blanking a comment keeps the length of the document', () => {
  expect(stripComments('a: b # why\n')).toBe('a: b      \n');
  expect(stripComments('"# kept"\n')).toBe('"# kept"\n');
  expect(stripComments('issue#1047\n')).toBe('issue#1047\n');
  expect(stripComments('a: b\n').length).toBe(5);
});

File: java/src/test/java/io/github/linkfoundation/linksnotation/CommentsTest.java
package io.github.linkfoundation.linksnotation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.stream.Collectors;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * Conformance tests for line comments.
 *
 * <p>https://github.com/link-foundation/links-notation/issues/301
 *
 * <p>{@code #} starts a comment that runs to the end of the line, unless it sits inside a token or
 * inside a delimited reference. Parsers read comments by default and can be told to treat {@code #}
 * as an ordinary character again. The table below is shared with the Rust, JavaScript, Python, Go,
 * C# and PHP suites, so a document written by one implementation reads the same in all of them.
 */
class CommentsTest {

  private Parser parser;

  @BeforeEach
  void setUp() {
    parser = new Parser();
  }

  /** Render a parsed node unambiguously: every reference is wrapped in angle brackets. */
  private static String render(Link node) {
    if (node.getValues() == null || node.getValues().isEmpty()) {
      return "<" + (node.getId() == null ? "" : node.getId()) + ">";
    }
    String head = node.getId() == null ? "" : "<" + node.getId() + ">: ";
    return "("
        + head
        + node.getValues().stream().map(CommentsTest::render).collect(Collectors.joining(" "))
        + ")";
  }

  private static String renderedWith(Parser parser, String source) throws ParseException {
    List<Link> links = parser.parse(source);
    return links.stream().map(CommentsTest::render).collect(Collectors.joining("\n"));
  }

  private String rendered(String source) throws ParseException {
    return renderedWith(parser, source);
  }

  private void assertParsesAs(String source, String expected) throws ParseException {
    assertEquals(expected, rendered(source), "Parsing " + source);
  }

  @Test
  void aLineThatStartsWithAHashIsAComment() throws ParseException {
    assertParsesAs("# a b\n", "");
  }

  @Test
  void aCommentMayHoldAColon() throws ParseException {
    // The document from #301: prose with a colon used to be read as a link.
    assertParsesAs("# a: b\n", "");
  }

  @Test
  void aCommentMayHoldAnythingAtAll() throws ParseException {
    assertParsesAs("# ) : ( \" ' ` #\n", "");
  }

  @Test
  void aCommentEndsAtTheEndOfItsLine() throws ParseException {
    assertParsesAs("# note\na: b\n", "(<a>: <b>)");
  }

  @Test
  void aCommentMayFollowALink() throws ParseException {
    assertParsesAs("a: b # why\n", "(<a>: <b>)");
  }

  @Test
  void aCommentMayFollowAGroup() throws ParseException {
    assertParsesAs("(a b) # why\n", "(<a> <b>)");
  }

  @Test
  void aCommentNeedsNoClosingNewline() throws ParseException {
    assertParsesAs("a: b # why", "(<a>: <b>)");
  }

  @Test
  void aCommentLineInsideAnIndentedBlockIsSkipped() throws ParseException {
    assertEquals(
        rendered("parent\n  child\n"), rendered("parent\n  # what the child is for\n  child\n"));
  }

  @Test
  void aCommentLineInsideAGroupIsSkipped() throws ParseException {
    assertEquals(rendered("(\n  a\n  b\n)\n"), rendered("(\n  a\n  # why\n  b\n)\n"));
  }

  @Test
  void aLineOfSpacesSeparatesLinksTheWayAnEmptyLineDoes() throws ParseException {
    // Blanking a comment leaves a line of spaces behind, so such a line has to
    // read the way an empty line does.
    assertEquals(rendered("a\n\nb\n"), rendered("a\n   \nb\n"));
  }

  @Test
  void aDocumentOfCommentsAloneHoldsNoLinks() throws ParseException {
    assertParsesAs("# one\n# two\n", "");
  }

  @Test
  void aHashInsideATokenIsAnOrdinaryCharacter() throws ParseException {
    assertParsesAs("issue#1047\n", "(<issue#1047>)");
  }

  @Test
  void aHashThatOpensATokenIsAnOrdinaryCharacter() throws ParseException {
    assertParsesAs("a: b#c\n", "(<a>: <b#c>)");
  }

  @Test
  void aHashInsideADelimitedReferenceIsContent() throws ParseException {
    assertParsesAs("\"# not a comment\" a\n", "(<# not a comment> <a>)");
  }

  @Test
  void aCommentMayFollowADelimitedReference() throws ParseException {
    assertParsesAs("\"a\" # why\n", "(<a>)");
  }

  @Test
  void aHashInsideAMultiLineDelimitedReferenceIsContent() throws ParseException {
    assertParsesAs("\"a # b\nc\" d\n", "(<a # b\nc> <d>)");
  }

  @Test
  void commentsAreOnByDefault() throws ParseException {
    assertParsesAs("# a b\n", "");
  }

  @Test
  void aParserWithoutCommentsKeepsTheHash() throws ParseException {
    assertEquals("(<#> <a> <b>)", renderedWith(new Parser(false), "# a b\n"));
  }

  @Test
  void blankingACommentKeepsTheLengthOfTheDocument() {
    assertEquals("a: b      \n", Comments.stripComments("a: b # why\n"));
    assertEquals("\"# kept\"\n", Comments.stripComments("\"# kept\"\n"));
    assertEquals("issue#1047\n", Comments.stripComments("issue#1047\n"));
    assertTrue(Comments.stripComments("# why\n").startsWith("     "));
  }

  @Test
  void aReferenceThatBeginsWithAHashIsWrittenQuoted() throws ParseException {
    // Without the quotes the document would read as `a` followed by a comment.
    String document =
        Link.formatLinks(List.of(new Link(null, List.of(new Link("a"), new Link("#tag")))));

    assertEquals("(a '#tag')", document);
    assertEquals("(<a> <#tag>)", rendered(document));
  }

  @Test
  void aHashThatCannotOpenACommentIsLeftUnquoted() {
    assertEquals("issue#1047", Link.escapeReference("issue#1047"));
    assertEquals("'#'", Link.escapeReference("#"));
    assertEquals("'#ff0000'", Link.escapeReference("#ff0000"));
  }
}
